### PR TITLE
feat: Expose connection params of the native client through the data service COMPASS-4517

### PR DESCRIPTION
**Pull request by gribnoysup**
## Description

Follow-up of https://github.com/mongodb-js/connection-model/pull/332 that now exposees mongo client connection params through the public `getConnectionParams` method of `DataService`

### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context

See https://github.com/mongodb-js/connection-model/pull/332

- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Dependents

- https://github.com/mongodb-js/connection-model/pull/332

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Review by rose-m — APPROVED** [review]
LGTM 👍

**Comment by rose-m** [comment]
nit regarding PR title: I think the issue should go at the end of the title - so `feat: Expose connection params of the native client through the data service COMPASS-4517`? 🤔

File: lib/data-service.js [hunk 1]
```diff
@@ -27,6 +27,10 @@ class DataService extends EventEmitter {
       .on('topologyDescriptionChanged', (evt) => this.emit('topologyDescriptionChanged', evt));
   }
 
+  getConnectionParams() {
```
Comment (alenakhineika) [inline lib/data-service.js:30]: How connectionParams are different from connectionOptions?

Comment (gribnoysup) [reply lib/data-service.js:30]: Are you referring to something specific in `data-service` when you say `connectionOptions`? I'm not sure what they are

File: test/native-client.test.js [hunk 2]
```diff
@@ -40,12 +40,23 @@ describe('NativeClient', function() {
 
       /*
        * pretends to be a connection-model providing every function call
-       * required in NativeClient#connect, but returns topology of our choice
+       * required in NativeClient#connect, but returns topology and connection
+       * params of our choice
        */
-      function mockedConnectionModel(topologyDescription) {
+      function mockedConnectionModel(topologyDescription, connectionOptions) {
```
Comment (alenakhineika) [inline test/native-client.test.js:46]: Here you pass connectionOptions.

Comment (gribnoysup) [reply test/native-client.test.js:46]: Ah, thanks for pointing that out! Yeah, this was not done for any specific reason, they are the same, and being more consistent in the name would definitely will be better. Will update it all to be `connectionOptions` 👍

**Comment by gribnoysup** [comment]
> nit regarding PR title: I think the issue should go at the end of the title - so feat: Expose connection params of the native client through the data service COMPASS-4517? 🤔

@rose-m is this something we have documented somewhere? I don't have a strong preference, but kinda makes sense for me for ticket to be the scope of the commit and github auto-highlight for jira tickets feature seems to work regardless

File: lib/data-service.js [hunk 3]
```diff
@@ -27,6 +27,10 @@ class DataService extends EventEmitter {
...
+    return this.client.connectionParams;
```
Comment (alenakhineika) [inline lib/data-service.js:31]: if this.client is a connection model, there are no connectionParams, only connectionOptions.

**Comment by alenakhineika** [comment]
For which task do you need these params? I am also not sure if data service should return them and not request them directly from client.

**Comment by rose-m** [comment]
@gribnoysup That's here: https://github.com/10gen/compass-internal-docs/blob/0d119fc19769cb14fb56ab2071205f2a90a77afb/technical/compass/how-to-work-on-compass.md#commit-message-formatting 😊

**Comment by gribnoysup** [comment]
@alenakhineika those are not exactly the same connection options that can be accessed directly from `connection-model`, those are processed connection options that are passed to the mongo client, we exposed them in https://github.com/mongodb-js/connection-model/pull/332 in `connection-model`.

The reason to expose them is part of the scope about moving shell runtime in Compass to the separate thread to unblock the UI. If we want to move it to a separate thread, we would not be able to pass mongo client to the repl directly anymore, instead we need a way to exactly replicate the mongo client connection in this new thread and for that we wanted a serializable version of connection options that `connection-model` used to connect to the server.

To avoid duplicating all the logic that `connection-model` applies to the connection options before passing them to the mongo client, we decided that it would be better to expose processed options on the `native-client` and made them available through the `data-service`.

As for exposing them through `data-service`, @mcasimir was saying that `data-service` is the public interface, while `native-client` is the internal implementation, and even though we access stuff directly from it sometimes, it should be better to expose everything that needs to be public through the `data-service`, so I just followed his guidance here. As an example of that `data-service` seems to proxy all `native-client` events without any additional processing and I think it's for the same reason of `data-service` being the public interface for `native-client`.

Does it gives more context on why it's implemented that way? Do you think we should approach it differently? I'm also happy to go through the "Background shell processing" tech design together and walk you through our investigation for the approach in detail if you feel like it, just ping me! 😃

**Comment by alenakhineika** [comment]
In this case yeah it makes sense. Just replace connectionParams with connectionOptions then.

**Comment by gribnoysup** [comment]
Changed the method name in b0d01e3. Will merge and release it now, thanks for the reviews!

**Pull request merged**